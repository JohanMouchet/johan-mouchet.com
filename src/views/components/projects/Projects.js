// @flow

import * as React from "react";
import cx from "classnames";
import "./Projects.scss";
import { ReactComponent as IconAward } from "assets/images/icons/award.svg";
import pluralize from "utils/pluralize";
import parse from "utils/parse";

type Props = {
  projects: [
    {
      highlight?: boolean,
      name: string,
      tagline?: string,
      url: string,
      lede: string,
      thumbnailPath: string,
      awards?: [
        {
          name: string,
          url: string,
          category: string,
        }
      ],
      features?: Array<string>,
      architecture?: Array<string>,
      libraries?: Array<string>,
    }
  ],
};

const Projects = ({ projects }: Props) => {
  const renderThumbnail = (project) => (
    <a
      href={project.url}
      target="_blank"
      rel="noopener noreferrer"
      className="c-project__link"
    >
      <img
        className="c-project__thumbnail"
        src={project.thumbnailPath}
        alt={project.name}
        loading="lazy"
      />
      <div
        href={project.url}
        target="_blank"
        rel="noopener noreferrer"
        className="c-project__overlay"
      >
        <h4 className="c-project__title">{project.name}</h4>
        <span className="c-project__line"></span>
        {project.tagline && (
          <span className="c-project__tagline">{project.tagline}</span>
        )}
      </div>
    </a>
  );

  const renderContent = (project, highlighted) => (
    <>
      {project.lede && (
        <p className={cx("c-project__lede", highlighted && "u-vr--top-0-@xs")}>
          {parse(project.lede)}
        </p>
      )}

      {project.awards && (
        <ul
          className={cx("c-project__awards", highlighted && "u-vr--top-0-@xs")}
        >
          {project.awards.map((award) => (
            <li className="c-project__award" key={award.name}>
              <a className="c-project__award-icon" href={award.url}>
                <IconAward />
              </a>
              <span className="c-project__award-name">
                <em>{award.name}</em>
                {award.category && ": " + award.category}
              </span>
            </li>
          ))}
        </ul>
      )}

      {(project.features || project.architecture || project.libraries) && (
        <ul
          className={cx(
            "c-project__details",
            highlighted && !project.awards && "u-vr--top-0-@xs"
          )}
        >
          {project.features && [
            <li className="c-project__details-section" key="features">
              <span className="c-project__details-heading">
                {pluralize("Feature", project.features.length)}
              </span>
              <ul className="c-project__detail-list">
                {project.features.map((detail) => (
                  <li className="c-project__detail" key={parse(detail)}>
                    {parse(detail)}
                  </li>
                ))}
              </ul>
            </li>,
          ]}

          {project.architecture && [
            <li className="c-project__details-section" key="architecture">
              <span className="c-project__details-heading">Architecture</span>
              <ul className="c-project__detail-list">
                {project.architecture.map((detail) => (
                  <li className="c-project__detail" key={parse(detail)}>
                    {parse(detail)}
                  </li>
                ))}
              </ul>
            </li>,
          ]}

          {project.libraries && [
            <li className="c-project__details-section" key="libraries">
              <span className="c-project__details-heading">
                {pluralize("Library", project.features.length)}
              </span>
              <ul className="c-project__detail-list">
                {project.libraries.map((detail) => (
                  <li className="c-project__detail" key={parse(detail)}>
                    {parse(detail)}
                  </li>
                ))}
              </ul>
            </li>,
          ]}
        </ul>
      )}
    </>
  );

  return !projects ? null : (
    <div className="c-projects">
      <div className="grid">
        {projects.map((project) =>
          project.highlight ? (
            <div
              className="cell cell--12-@xs"
              key={project.name + project.tagline}
            >
              <div className="c-project c-project--single">
                <div className="grid">
                  <div className="cell cell--12-@xs cell--6-@sm cell--4-@lg">
                    {renderThumbnail(project)}
                  </div>
                  <div className="cell cell--12-@xs cell--6-@sm cell--8-@lg">
                    {renderContent(project, true)}
                  </div>
                </div>
              </div>
            </div>
          ) : (
            <div
              className={cx("cell cell--12-@xs cell--6-@sm", {
                "cell--4-@lg": project.lede,
                "cell--3-@lg": !project.lede,
              })}
              key={project.name + project.tagline}
            >
              <div className="c-project">
                {renderThumbnail(project)}
                {renderContent(project)}
              </div>
            </div>
          )
        )}
      </div>
    </div>
  );
};

export default Projects;
