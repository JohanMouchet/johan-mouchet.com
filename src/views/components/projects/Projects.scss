@import "assets/styles/settings/_settings";
@import "assets/styles/tools/_tools";

.c-project {
  $root: &;

  &__link {
    display: block;
    max-width: 430px;
    margin-bottom: t-vertical-rhythm(1);
    position: relative;
    overflow: hidden;

    &:hover {
      #{$root}__thumbnail {
        filter: grayscale(0);
      }

      #{$root}__overlay {
        text-decoration: none;
        opacity: 0;

        /* stylelint-disable-next-line max-nesting-depth */
        &::before {
          transform: scaleX(0);
        }
      }
    }
  }

  &__thumbnail {
    display: block;
    width: 100%;
    border: 1px solid $s-color-gray-lighter;
    filter: grayscale(1);
    transition: filter 0.2s ease-out;
  }

  &__overlay {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    color: $s-color-white;
    text-align: center;
    transition: opacity 0.2s ease-out;

    &::before {
      content: "";
      position: absolute;
      top: 0;
      left: 0;
      height: 100%;
      width: 100%;
      background-image: t-ui(pattern, dots),
        linear-gradient(
          rgba($s-color-primary-lighter, 0.7),
          rgba($s-color-primary-darker, 0.9)
        );
      transform-origin: 100% 50%;
      transition: transform 0.2s ease-out;
    }

    > * {
      position: relative;
    }
  }

  &__title {
    margin: 0;
    font-style: italic;
    font-weight: 700;
    color: $s-color-secondary;
  }

  &__line {
    width: 30%;
    border-bottom: 1px solid;
    margin-top: t-vertical-rhythm(0.3);
    margin-bottom: t-vertical-rhythm(0.3);
  }

  &__tagline {
    margin: 0;
    font-size: t-modular-scale(-1);
  }

  &__details-heading {
    font-weight: 500;
  }

  &__details-list {
    margin-top: 0;
  }

  &__awards {
    padding-left: 0;
    list-style-type: none;
    position: relative;
  }

  &__award {
    display: flex;

    &-icon {
      flex: 0 1 1.2em;
      align-self: flex-start;
      margin-right: 0.3em;
    }

    &-name {
      flex: 1 1 0%;
    }

    svg {
      height: 1.2em;
      width: 1.2em;
      fill: currentColor;
      vertical-align: middle;
    }
  }

  &--highlighted {
    margin-bottom: t-vertical-rhythm(1);
    padding: t-vertical-rhythm(1) t-vertical-rhythm(1) 0;
    border: 1.99px solid $s-color-secondary-lighter;
    border-radius: 6px;

    @include t-wekkit-hdpi-small-dimension;
  }
}
