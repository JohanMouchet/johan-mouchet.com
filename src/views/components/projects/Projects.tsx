import * as React from "react";
import cx from "classnames";
import "./Projects.scss";
import { Details } from "views/objects";
import pluralize from "utils/pluralize";
import parse from "utils/parse";

type Thumbnail = {
  name: string;
  tagline?: string;
  url: string;
  lede: string;
  thumbnailSrc: string;
};

type Content = {
  highlight?: boolean;
  achievements?: string[];
  architecture?: string[];
  libraries?: string[];
};

type Props = {
  projects: (Thumbnail & Content)[];
};

const Projects: React.FC<Props> = ({ projects }) => {
  const renderThumbnail = (project: Thumbnail) => (
    <>
      <a href={project.url} className="c-project__link">
        <img
          className="c-project__thumbnail"
          src={project.thumbnailSrc}
          alt={project.name}
          loading="lazy"
        />
        <a href={project.url} className="c-project__overlay">
          <h4 className="c-project__title">{project.name}</h4>
          <span className="c-project__line"></span>
          {project.tagline && (
            <span className="c-project__tagline">{project.tagline}</span>
          )}
        </a>
      </a>

      {project.lede && <p className="c-project__lede">{parse(project.lede)}</p>}
    </>
  );

  const renderContent = (project: Content) => (
    <>
      {project.achievements && (
        <div className="c-project__detail">
          <b className="c-project__detail-heading">
            Key {pluralize("Achievement", project.achievements.length)}
          </b>
          <ul className="c-project__detail-list">
            {project.achievements.map((detail) => (
              <li key={String(parse(detail))}>{parse(detail)}</li>
            ))}
          </ul>
        </div>
      )}

      {project.architecture && (
        <div className="c-project__detail">
          <b className="c-project__detail-heading">Architecture</b>
          <p className="c-project__detail-list">
            {parse(project.architecture.join(", "))}
          </p>
        </div>
      )}

      {project.libraries && (
        <div className="c-project__detail">
          <b className="c-project__detail-heading">
            {pluralize("Library", project.libraries.length)}
          </b>
          <p className="c-project__detail-list">
            {parse(project.libraries.join(", "))}
          </p>
        </div>
      )}
    </>
  );

  return !projects ? null : (
    <div className="c-projects">
      <div className="grid">
        {projects.map((project) =>
          project.highlight ? (
            <div
              className="cell cell--12-@xs"
              key={project.name + (project.tagline || "")}
            >
              <div className="c-project c-project--highlighted">
                <div className="grid">
                  <div className="cell cell--12-@xs cell--6-@sm cell--4-@lg">
                    {renderThumbnail(project)}
                  </div>
                  <div className="cell cell--12-@xs cell--6-@sm cell--8-@lg">
                    {renderContent(project)}
                  </div>
                </div>
              </div>
            </div>
          ) : (
            <div
              className={cx("cell cell--12-@xs cell--6-@sm", {
                "cell--4-@lg": project.lede,
                "cell--3-@lg": !project.lede,
              })}
              key={project.name + (project.tagline || "")}
            >
              <div className="c-project">
                {renderThumbnail(project)}
                {(project.achievements ||
                  project.architecture ||
                  project.libraries) && (
                  <Details summary="Details" variant="compact">
                    <div className="c-project__details">
                      {renderContent(project)}
                    </div>
                  </Details>
                )}
              </div>
            </div>
          )
        )}
      </div>
    </div>
  );
};

export default Projects;
