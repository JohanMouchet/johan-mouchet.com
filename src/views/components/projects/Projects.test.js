import React from "react";
import { shallow } from "enzyme";
import Projects from ".";

it("Renders <Projects /> without crashing", () => {
  shallow(<Projects />);
});
