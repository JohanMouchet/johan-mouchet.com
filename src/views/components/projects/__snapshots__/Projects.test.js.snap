// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <Projects /> without crashing 1`] = `""`;
